lib/utils/verify-signatures.js

```diff
@@ -0,0 +1,460 @@
+    this.getEdges(nodes, 'edgesOut')
+
+    const edges = Array.from(this.edges)
+    const start = process.hrtime.bigint()
+    const registries = this.findAllRegistryUrls(edges, this.npm.flatOptions)
+
+    // Prefetch and cache public keys from the used registries
+    await Promise.all(registries.map(async (registry) => this.getKeys({ registry })))
+    await Promise.all(edges.map(async (edge) => await this.getVerifiedInfo(edge)))
+
+    const end = process.hrtime.bigint()
+    const elapsed = end - start
+
+    // sort alphabetically
+    const invalid = Array.from(this.invalid).sort((a, b) => localeCompare(a.name, b.name))
+    const missing = Array.from(this.missing).sort((a, b) => localeCompare(a.name, b.name))
+
+    const verified = invalid.length === 0 && missing.length === 0
+
+    if (!verified) {
+      this.exitCode = 1
+    }
+
+    if (this.npm.config.get('json')) {
+      this.appendOutput(this.makeJSON({ invalid, missing }))
+    } else {
+      const timing = `audited ${edges.length} packages in ${Math.floor(Number(elapsed) / 1e9)}s`
+      const verifiedPrefix = verified ? 'verified signatures, ' : ''
+      this.appendOutput(`${verifiedPrefix}${timing}\n`)
+
+      if (missing.length) {
+        const msg = missing.length === 1 ?
+          `package has a ${chalk.bold(chalk.magenta('missing'))} registry signature` :
+          `packages have ${chalk.bold(chalk.magenta('missing'))} registry signatures`
+        this.appendOutput(
+          `${missing.length} ${msg} but the registry is ` +
+          `providing signing keys${this.npm.config.get('missing') ? ':\n' : ''}`
+        )
+        // TODO: This might not be the right option for this
+        if (this.npm.config.get('missing')) {
+          this.appendOutput(this.humanOutput(missing))
+        } else {
+          this.appendOutput(`  run \`npm audit signatures --missing\` for details`)
+        }
+      }
+
+      if (invalid.length) {
+        const msg = invalid.length === 1 ?
+          `package has an ${chalk.bold(chalk.red('invalid'))} registry signature` :
+          `packages have ${chalk.bold(chalk.red('invalid'))} registry signatures`
+        this.appendOutput(
+          `${missing.length ? '\n' : ''}${invalid.length} ${msg}:\n`
+        )
+        this.appendOutput(this.humanOutput(invalid))
+        const plural = invalid.length === 1 ? '' : 's'
+        this.appendOutput(
+          `\nSomeone might have tampered with the package${plural} ` +
+          `since being published on the registry (monster-in-the-middle attack)!\n`
+        )
+      }
+    }
+  }
+
+  findAllRegistryUrls (edges, opts) {
+    return Array.from(edges.reduce((p, edge) => {
+      let alias = false
+      try {
+        alias = npa(edge.spec).subSpec
+      } catch (err) {
+      }
+      const spec = npa(alias ? alias.name : edge.name)
+      p.add(fetch.pickRegistry(spec, opts))
+      return p
+    }, new Set()))
+  }
+
+  appendOutput (...args) {
+    args = [...args].flat()
```

Comment:
```suggestion
    args = args.flat()
```
the “rest” in the signature already makes it an array

an alternative would be for this entire method body to be `this.output.push(…args.flat())`